src/backend/distributed/executor/merge_executor.c

```diff
@@ -0,0 +1,422 @@
+													   work_mem);
+	TupleDesc tupleDescriptor = ScanStateGetTupleDescriptor(scanState);
+	TupleDestination *tupleDest =
+		CreateTupleStoreTupleDest(scanState->tuplestorestate, tupleDescriptor);
+	uint64 rowsMerged =
+		ExecuteTaskListIntoTupleDest(ROW_MODIFY_COMMUTATIVE, prunedTaskList,
+									 tupleDest, hasReturning);
+	executorState->es_processed = rowsMerged;
+}
+
+
+/*
+ * sourceResultPartitionColumnIndex collects all Join conditions from the
+ * ON clause and verifies if there is a join, either left or right, with
+ * the distribution column of the given target. Once a match is found, it
+ * returns the index of that match in the source's target list.
+ */
+static int
+sourceResultPartitionColumnIndex(Query *mergeQuery, List *sourceTargetList,
+								 CitusTableCacheEntry *targetRelation)
+{
+	/* Get all the equal(=) expressions from the ON clause */
+	List *mergeJoinConditionList = QualifierList(mergeQuery->jointree);
+
+	/*
+	 * Now that we have gathered all the _equal_ expressions in the list, we proceed
+	 * to match the one that joins with the distribution column of the target.
+	 */
+	Var *targetColumn = targetRelation->partitionColumn;
+	List *sourceJoinColumns = NIL;
+	Node *joinClause = NULL;
+	foreach_ptr(joinClause, mergeJoinConditionList)
+	{
+		Node *leftOperand;
+		Node *rightOperand;
+
+		if (!BinaryOpExpression((Expr *) joinClause, &leftOperand, &rightOperand))
+		{
+			continue;
+		}
+
+		if (equal(leftOperand, targetColumn) && IsA(rightOperand, Var))
+		{
+			sourceJoinColumns = lappend(sourceJoinColumns, rightOperand);
+		}
+		else if (equal(rightOperand, targetColumn) && IsA(leftOperand, Var))
+		{
+			sourceJoinColumns = lappend(sourceJoinColumns, leftOperand);
+		}
+	}
+
+	int partitionColumnIndex = -1;
+	if (sourceJoinColumns)
+	{
+		/* Use the Var that's being used in INSERT clause */
+		Var *sourceJoinColumn =
+			ErrorIfInsertNotMatchTargetDistributionColumn(
+				targetRelation->relationId, mergeQuery,
+				sourceJoinColumns);
+
+		/* If no INSERT action is present, we can use one of the Var */
+		if (!sourceJoinColumn)
+		{
+			sourceJoinColumn = linitial(sourceJoinColumns);
+		}
+
+		/*
+		 * Find the index in the source list that will be utilized
+		 * for colocation-redistribution.
+		 */
+		partitionColumnIndex =
+			DistributionColumnIndex(sourceTargetList, sourceJoinColumn);
+	}
+
+	if (partitionColumnIndex == -1)
+	{
+		ereport(ERROR, (errmsg("Missing required join condition between "
+							   "source and target's distribution column"),
+						errhint("Without a join condition on the target's "
+								"distribution column, the source rows "
+								"cannot be efficiently redistributed, and "
+								"the NOT-MATCHED condition cannot be evaluated "
+								"unambiguously. This can result in incorrect or "
+								"unexpected results when attempting to merge "
+								"tables in a distributed setting")));
```

Comment:
These checks are probably better suited for planner, otherwise the explain shows it properly but the actual execution fails.

```sql
explain MERGE INTO target USING source ON (source.key + 1 = target.key)
        WHEN MATCHED THEN UPDATE SET value = 15;
┌──────────────────────────────────────────────────────────────────────────────────────────┐
│                                        QUERY PLAN                                        │
├──────────────────────────────────────────────────────────────────────────────────────────┤
│ Custom Scan (Citus MERGE INTO ...)  (cost=0.00..0.00 rows=0 width=0)                     │
│   MERGE INTO target method: repartition                                                  │
│   ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=8)                │
│         Task Count: 4                                                                    │
│         Tasks Shown: One of 4                                                            │
│         ->  Task                                                                         │
│               Node: host=localhost port=9701 dbname=onderkalaci                          │
│               ->  Seq Scan on source_102228 source  (cost=0.00..32.60 rows=2260 width=8) │
└──────────────────────────────────────────────────────────────────────────────────────────┘
(8 rows)

[local] onderkalaci@onderkalaci:9700-12420=#
MERGE INTO target USING source ON (source.key + 1 = target.key)
        WHEN MATCHED THEN UPDATE SET value = 15;
ERROR:  Missing required join condition between source and target's distribution column
HINT:  Without a join condition on the target's distribution column, the source rows cannot be efficiently redistributed, and the NOT-MATCHED condition cannot be evaluated unambiguously. This can result in incorrect or unexpected results when attempting to merge tables in a distributed setting
```

in general, this also looks like a planning error anyway.

Reply:
@onderkalaci lot of comments are outdated, but I'm still looking at them (anything that's not marked as resolved). I'm giving priority to the functional bugs/crashes. I will address these too

Reply:
@onderkalaci It makes sense to fail it planning. Moved the check to planning phase and added the above testcase in regression suite.